core/include/vecmem/memory/choice_memory_resource.hpp

```diff
@@ -0,0 +1,55 @@
+/**
+ * VecMem project, part of the ACTS project (R&D line)
+ *
+ * (c) 2021 CERN for the benefit of the ACTS project
+ *
+ * Mozilla Public License Version 2.0
+ */
+
+#pragma once
+
+#include <cstddef>
+#include <functional>
+#include <map>
+#include <vector>
+
+#include "vecmem/memory/memory_resource.hpp"
+
+namespace vecmem {
+/**
+ * @brief This memory resource conditionally allocates memory. It is
+ * constructed with a function that determines which upstream resource to use.
+ *
+ * This resource can be used to construct complex conditional allocation
+ * schemes.
+ */
+class choice_memory_resource : public memory_resource {
+public:
+    /**
+     * @brief Constructs the choice memory resource.
+     *
+     * @param[in] upstreams The upstream memory resources to use.
+     * @param[in] decision The function which picks the upstream memory
+     * resource to use by index.
+     */
+    choice_memory_resource(
+        std::initializer_list<std::reference_wrapper<memory_resource>>&&
+            upstreams,
```

Comment:
Hmm, I disagree with you for two reasons. Firstly, `initializer_list` has a static length, while `vector` has a dynamic length. That's pretty obvious, but it means that the vector approach has more overhead than the initializer list approach. Granted, the difference is extremely small, but it's there. Secondly, using the vector approach relies on implicit conversion. That's not a problem per sé, but personally I prefer the explicit approach.

This also gives us access to `{}` constructor syntax for the coalescing resource (I think), which we wouldn't have otherwise (I think).

Maybe these links will help you make up your mind:

[1] https://stackoverflow.com/questions/14414832/why-use-initializer-list-instead-of-vector-in-parameters
[2] https://stackoverflow.com/questions/27753420/initializer-list-vs-vector

Reply:
The first and only thing that you do with this initializer_list is to give it to the `std::vector` variable that you have as a member in the class. So I'm really not sure where the upside to this syntax is. :confused:

In this setup you can only construct your `vecmem::choice_memory_resource` object with a static list of upstream resources. If the constructor received a vector, it would also become possible to construct such an object "with some algorithmic code". (Which would decide at runtime what memory resources are available, and set up an appropriate selection function on top of them.)

I do know for a fact that you can easily use code like the following:

```c++
struct some_type {
   some_type( const std::vector<float>& v1, const std::vector<int>& v2 );
};

some_type t( { 0.1, 0.2 }, { 1, 2, 3 } );
```

The whole point of STL containers being initialisable from initializer lists is to be able to do this automatically. I jumped on using `std::initializer_list` in this type of way in the past. But had to realise afterwards that (as I'm trying to lay out here) my code was not any better than if I just used a vector directly. Since I also always ended up storing the received elements in an internal vector anyway.

Reply:
Being able to construct vectors algorithmically is a _very_ good point. :thinking: